api/client/src/main/java/org/projectnessie/client/auth/oauth2/AuthorizationCodeFlow.java

```diff
@@ -0,0 +1,242 @@
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.io.UncheckedIOException;
+import java.io.UnsupportedEncodingException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.projectnessie.client.http.HttpRequest;
+import org.projectnessie.client.http.HttpResponse;
+import org.projectnessie.client.http.impl.HttpUtils;
+import org.projectnessie.client.http.impl.UriBuilder;
+
+class AuthorizationCodeFlow implements AutoCloseable {
+
+  public static final String CONTEXT_PATH = "/nessie-client/auth";
+  public static final String MSG_PREFIX = "[nessie-oauth2-client] ";
+
+  private static final org.slf4j.Logger LOGGER =
+      org.slf4j.LoggerFactory.getLogger(AuthorizationCodeFlow.class);
+
+  private static final String HTML_TEMPLATE_OK =
+      "<html><body><h1>Authentication successful</h1><p>You can close this page now.</p></body></html>";
+  private static final String HTML_TEMPLATE_FAILED =
+      "<html><body><h1>Authentication failed</h1><p>Could not obtain access token: %s</p></body></html>";
+
+  private static final int STATE_LENGTH = 16;
+
+  private final OAuth2Client client;
+  private final PrintStream console;
+  private final String state;
+  private final HttpServer server;
+  private final String redirectUri;
+  private final URI authorizationUri;
+
+  private final Duration flowTimeout = Duration.ofMinutes(5);
+
+  private final CompletableFuture<HttpExchange> requestFuture = new CompletableFuture<>();
+  private final CompletableFuture<Void> responseFuture = new CompletableFuture<>();
+  private final CompletableFuture<Tokens> tokensFuture;
+  private final CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+
+  AuthorizationCodeFlow(OAuth2Client client) {
+    this(client, System.out);
+  }
+
+  /** Constructor used for testing. */
+  AuthorizationCodeFlow(OAuth2Client client, PrintStream console) {
+    this.client = client;
+    this.console = console;
+    // Close the server when the close future completes, but wait for the response to be sent first,
+    // so that the browser does not show "Connection closed" to the user.
+    CompletableFuture.allOf(closeFuture, responseFuture).thenRun(this::doClose);
+    tokensFuture =
+        requestFuture.thenApply(this::extractAuthorizationCode).thenApply(this::fetchNewTokens);
+    server = createServer(client.authServerPort, this::doRequest);
+    state = OAuth2Utils.randomAlphaNumString(STATE_LENGTH);
+    redirectUri =
+        String.format("http://localhost:%d%s", server.getAddress().getPort(), CONTEXT_PATH);
```

Comment:
Should we allow the user to inject the authrorization code into the Nessie Client manually, via its configuration?

Reply:
> How can it work with `localhost` in real life deployments? I suppose we can expect the Nessie Client's STDOUT to be available to the user, but can we assume that the browser runs on the same host as the client?.. I'm not sure.

The authorization code flow _requires_ a human user and a browser to be available. From the RFC:

> Since this is a redirection-based flow, the client must be capable of
   interacting with the resource owner's user-agent (typically a web
   browser) and capable of receiving incoming requests (via redirection)
   from the authorization server.

So this specific flow is only an option in the situation where a human has some sort of terminal session opened, e.g. Spark SQL, and also a web browser available (even curl is OK, if nothing else). It's not an option for situations where the Nessie client is running in a non-interactive way.

Reply:
> Should we allow the user to inject the authrorization code into the Nessie Client manually, via its configuration?

It's more complicated than that: the Nessie client must generate a random challenge, called "state", and pass it to the authorization server as a query param. The server MUST reply with a redirect URI containing the state, plus a code that it generates. Then finally, the Nessie client must validate the received state, and invoke the tokens endpoint with the received code. 

Also, the code is supposed to change at every authentication attempt.

Reply:
Should we perhaps document that, in case Nessie Client is not co-located with the browser, the user is supposed to forward the "embedded" port from the host where the browser runs to the host where the Nessie Client runs?

Reply:
Yes, good idea.

Reply:
In hindsight, I would like to propose a different option: for the port forwarding to work, the Nessie client web server must listen to all interfaces. That is imho a security threat, and we should instead limit that web server, by default, to only the loopback address. I therefore would like to restrict the web server to localhost in this PR, then in a follow-up PR, we would explore the port forwarding option and likely introduce a new config option to allow the web server to listen on all interfaces. Wdyt?

Reply:
(or better yet, an option to specify the exact IP address to bind to.)